{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis tools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Swan has several analysis options to use. \n",
    "* [Differential gene expression](#deg)\n",
    "* [Differential transcript expression](#det)\n",
    "* [Isoform switching](#is)\n",
    "* [Exon skipping and intron retention](#es_ir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Graph from data/swan.p loaded\n"
     ]
    }
   ],
   "source": [
    "import swan_vis as swan\n",
    "\n",
    "sg = swan.SwanGraph('data/swan.p')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a name=\"deg\"></a>Differential gene expression tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Differential gene expression testing in Swan is implemented via [diffxpy](https://github.com/theislab/diffxpy). To run the test, first partition the datasets that you have added to your SwanGraph into biological replicates. Then, use this grouping to run the differential expression test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Transforming to str index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training location model: False\n",
      "training scale model: True\n",
      "iter   0: ll=76839008.205068\n",
      "iter   1: ll=76839008.205068, converged: 0.00% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter   2: ll=3994961.414960, converged: 1.20% (loc: 1.20%, scale update: True), in 112.66sec\n",
      "iter   3: ll=3994961.414960, converged: 1.20% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter   4: ll=2972738.740537, converged: 15.96% (loc: 15.96%, scale update: True), in 85.66sec\n",
      "iter   5: ll=2972738.740537, converged: 15.96% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter   6: ll=2919018.652501, converged: 21.45% (loc: 21.45%, scale update: True), in 79.18sec\n",
      "iter   7: ll=2919018.652501, converged: 21.45% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter   8: ll=2746114.113052, converged: 24.10% (loc: 24.10%, scale update: True), in 74.59sec\n",
      "iter   9: ll=2746114.113052, converged: 24.10% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter  10: ll=2745964.853568, converged: 99.88% (loc: 99.88%, scale update: True), in 66.71sec\n",
      "iter  11: ll=2745964.853568, converged: 99.88% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter  12: ll=2745900.750399, converged: 99.96% (loc: 99.96%, scale update: True), in 2.11sec\n",
      "iter  13: ll=2745900.750399, converged: 99.96% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter  14: ll=2745893.960127, converged: 99.99% (loc: 99.99%, scale update: True), in 2.03sec\n",
      "iter  15: ll=2745893.960127, converged: 99.99% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter  16: ll=2745893.960127, converged: 100.00% (loc: 100.00%, scale update: True), in 2.16sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/dask/array/core.py:2622: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  size = (limit / dtype.itemsize / largest_block) ** (1 / len(autos))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>pval</th>\n",
       "      <th>qval</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>mean</th>\n",
       "      <th>zero_mean</th>\n",
       "      <th>grad</th>\n",
       "      <th>coef_mle</th>\n",
       "      <th>coef_sd</th>\n",
       "      <th>ll</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93003</th>\n",
       "      <td>ENSG00000158874.11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.798466</td>\n",
       "      <td>7203.040430</td>\n",
       "      <td>False</td>\n",
       "      <td>8.236582e-07</td>\n",
       "      <td>-9.798466</td>\n",
       "      <td>0.597998</td>\n",
       "      <td>-21.964623</td>\n",
       "      <td>APOA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99184</th>\n",
       "      <td>ENSG00000163631.16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.657541</td>\n",
       "      <td>6256.314453</td>\n",
       "      <td>False</td>\n",
       "      <td>2.962665e-07</td>\n",
       "      <td>-9.657541</td>\n",
       "      <td>0.583054</td>\n",
       "      <td>-20.338973</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79865</th>\n",
       "      <td>ENSG00000145192.12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.320415</td>\n",
       "      <td>4466.045703</td>\n",
       "      <td>False</td>\n",
       "      <td>4.583478e-08</td>\n",
       "      <td>-9.320415</td>\n",
       "      <td>0.591606</td>\n",
       "      <td>-20.613694</td>\n",
       "      <td>AHSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4810</th>\n",
       "      <td>ENSG00000026025.15</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.226718</td>\n",
       "      <td>9419.225325</td>\n",
       "      <td>False</td>\n",
       "      <td>1.069652e-06</td>\n",
       "      <td>9.226718</td>\n",
       "      <td>0.569277</td>\n",
       "      <td>-24.014993</td>\n",
       "      <td>VIM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139314</th>\n",
       "      <td>ENSG00000197249.13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.106863</td>\n",
       "      <td>5051.564227</td>\n",
       "      <td>False</td>\n",
       "      <td>1.033368e-06</td>\n",
       "      <td>-9.106863</td>\n",
       "      <td>0.496072</td>\n",
       "      <td>-20.785889</td>\n",
       "      <td>SERPINA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176368</th>\n",
       "      <td>ENSG00000242968.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.912871</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>AC096992.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176367</th>\n",
       "      <td>ENSG00000242963.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.912871</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>AC026336.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176366</th>\n",
       "      <td>ENSG00000242960.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.912871</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>FTH1P23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176365</th>\n",
       "      <td>ENSG00000242958.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.912871</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>AC040975.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163669</th>\n",
       "      <td>ENSG00000230383.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.912871</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>AC009245.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58906 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       gid  pval  qval    log2fc         mean  zero_mean  \\\n",
       "93003   ENSG00000158874.11   0.0   0.0 -9.798466  7203.040430      False   \n",
       "99184   ENSG00000163631.16   0.0   0.0 -9.657541  6256.314453      False   \n",
       "79865   ENSG00000145192.12   0.0   0.0 -9.320415  4466.045703      False   \n",
       "4810    ENSG00000026025.15   0.0   0.0  9.226718  9419.225325      False   \n",
       "139314  ENSG00000197249.13   0.0   0.0 -9.106863  5051.564227      False   \n",
       "...                    ...   ...   ...       ...          ...        ...   \n",
       "176368   ENSG00000242968.1   1.0   1.0  0.000000     1.000000      False   \n",
       "176367   ENSG00000242963.1   1.0   1.0  0.000000     1.000000      False   \n",
       "176366   ENSG00000242960.1   1.0   1.0  0.000000     1.000000      False   \n",
       "176365   ENSG00000242958.1   1.0   1.0  0.000000     1.000000      False   \n",
       "163669   ENSG00000230383.1   1.0   1.0  0.000000     1.000000      False   \n",
       "\n",
       "                grad  coef_mle   coef_sd         ll       gname  \n",
       "93003   8.236582e-07 -9.798466  0.597998 -21.964623       APOA2  \n",
       "99184   2.962665e-07 -9.657541  0.583054 -20.338973         ALB  \n",
       "79865   4.583478e-08 -9.320415  0.591606 -20.613694        AHSG  \n",
       "4810    1.069652e-06  9.226718  0.569277 -24.014993         VIM  \n",
       "139314  1.033368e-06 -9.106863  0.496072 -20.785889    SERPINA1  \n",
       "...              ...       ...       ...        ...         ...  \n",
       "176368  0.000000e+00  0.000000  0.912871   0.000000  AC096992.1  \n",
       "176367  0.000000e+00  0.000000  0.912871   0.000000  AC026336.1  \n",
       "176366  0.000000e+00  0.000000  0.912871   0.000000     FTH1P23  \n",
       "176365  0.000000e+00  0.000000  0.912871   0.000000  AC040975.1  \n",
       "163669  0.000000e+00  0.000000  0.912871   0.000000  AC009245.1  \n",
       "\n",
       "[58906 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_groups = [['HepG2_1','HepG2_2'],['HFFc6_1','HFFc6_2','HFFc6_3']]\n",
    "sg.de_gene_test(dataset_groups)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results of this test are stored in `sg.deg_test` so they can be accessed later as follows. Test results will also be stored if `save_graph()` is run again so the user can easily load the results up."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>pval</th>\n",
       "      <th>qval</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>mean</th>\n",
       "      <th>zero_mean</th>\n",
       "      <th>grad</th>\n",
       "      <th>coef_mle</th>\n",
       "      <th>coef_sd</th>\n",
       "      <th>ll</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93003</th>\n",
       "      <td>ENSG00000158874.11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.798466</td>\n",
       "      <td>7203.040430</td>\n",
       "      <td>False</td>\n",
       "      <td>8.236582e-07</td>\n",
       "      <td>-9.798466</td>\n",
       "      <td>0.597998</td>\n",
       "      <td>-21.964623</td>\n",
       "      <td>APOA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99184</th>\n",
       "      <td>ENSG00000163631.16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.657541</td>\n",
       "      <td>6256.314453</td>\n",
       "      <td>False</td>\n",
       "      <td>2.962665e-07</td>\n",
       "      <td>-9.657541</td>\n",
       "      <td>0.583054</td>\n",
       "      <td>-20.338973</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79865</th>\n",
       "      <td>ENSG00000145192.12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.320415</td>\n",
       "      <td>4466.045703</td>\n",
       "      <td>False</td>\n",
       "      <td>4.583478e-08</td>\n",
       "      <td>-9.320415</td>\n",
       "      <td>0.591606</td>\n",
       "      <td>-20.613694</td>\n",
       "      <td>AHSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4810</th>\n",
       "      <td>ENSG00000026025.15</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.226718</td>\n",
       "      <td>9419.225325</td>\n",
       "      <td>False</td>\n",
       "      <td>1.069652e-06</td>\n",
       "      <td>9.226718</td>\n",
       "      <td>0.569277</td>\n",
       "      <td>-24.014993</td>\n",
       "      <td>VIM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139314</th>\n",
       "      <td>ENSG00000197249.13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.106863</td>\n",
       "      <td>5051.564227</td>\n",
       "      <td>False</td>\n",
       "      <td>1.033368e-06</td>\n",
       "      <td>-9.106863</td>\n",
       "      <td>0.496072</td>\n",
       "      <td>-20.785889</td>\n",
       "      <td>SERPINA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       gid  pval  qval    log2fc         mean  zero_mean  \\\n",
       "93003   ENSG00000158874.11   0.0   0.0 -9.798466  7203.040430      False   \n",
       "99184   ENSG00000163631.16   0.0   0.0 -9.657541  6256.314453      False   \n",
       "79865   ENSG00000145192.12   0.0   0.0 -9.320415  4466.045703      False   \n",
       "4810    ENSG00000026025.15   0.0   0.0  9.226718  9419.225325      False   \n",
       "139314  ENSG00000197249.13   0.0   0.0 -9.106863  5051.564227      False   \n",
       "\n",
       "                grad  coef_mle   coef_sd         ll     gname  \n",
       "93003   8.236582e-07 -9.798466  0.597998 -21.964623     APOA2  \n",
       "99184   2.962665e-07 -9.657541  0.583054 -20.338973       ALB  \n",
       "79865   4.583478e-08 -9.320415  0.591606 -20.613694      AHSG  \n",
       "4810    1.069652e-06  9.226718  0.569277 -24.014993       VIM  \n",
       "139314  1.033368e-06 -9.106863  0.496072 -20.785889  SERPINA1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.deg_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Swan can also automatically subset the test summary table to pull out genes that pass a certain significance threshold. These genes can be directly passed into Swan's gene plotting functions, `gen_report()` or `plot_graph()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ids, gene_summary = sg.get_de_genes(q=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['APOA2', 'ALB', 'AHSG', 'VIM', 'SERPINA1']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>pval</th>\n",
       "      <th>qval</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>mean</th>\n",
       "      <th>zero_mean</th>\n",
       "      <th>grad</th>\n",
       "      <th>coef_mle</th>\n",
       "      <th>coef_sd</th>\n",
       "      <th>ll</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93003</th>\n",
       "      <td>ENSG00000158874.11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.798466</td>\n",
       "      <td>7203.040430</td>\n",
       "      <td>False</td>\n",
       "      <td>8.236582e-07</td>\n",
       "      <td>-9.798466</td>\n",
       "      <td>0.597998</td>\n",
       "      <td>-21.964623</td>\n",
       "      <td>APOA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99184</th>\n",
       "      <td>ENSG00000163631.16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.657541</td>\n",
       "      <td>6256.314453</td>\n",
       "      <td>False</td>\n",
       "      <td>2.962665e-07</td>\n",
       "      <td>-9.657541</td>\n",
       "      <td>0.583054</td>\n",
       "      <td>-20.338973</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79865</th>\n",
       "      <td>ENSG00000145192.12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.320415</td>\n",
       "      <td>4466.045703</td>\n",
       "      <td>False</td>\n",
       "      <td>4.583478e-08</td>\n",
       "      <td>-9.320415</td>\n",
       "      <td>0.591606</td>\n",
       "      <td>-20.613694</td>\n",
       "      <td>AHSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4810</th>\n",
       "      <td>ENSG00000026025.15</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.226718</td>\n",
       "      <td>9419.225325</td>\n",
       "      <td>False</td>\n",
       "      <td>1.069652e-06</td>\n",
       "      <td>9.226718</td>\n",
       "      <td>0.569277</td>\n",
       "      <td>-24.014993</td>\n",
       "      <td>VIM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139314</th>\n",
       "      <td>ENSG00000197249.13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.106863</td>\n",
       "      <td>5051.564227</td>\n",
       "      <td>False</td>\n",
       "      <td>1.033368e-06</td>\n",
       "      <td>-9.106863</td>\n",
       "      <td>0.496072</td>\n",
       "      <td>-20.785889</td>\n",
       "      <td>SERPINA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       gid  pval  qval    log2fc         mean  zero_mean  \\\n",
       "93003   ENSG00000158874.11   0.0   0.0 -9.798466  7203.040430      False   \n",
       "99184   ENSG00000163631.16   0.0   0.0 -9.657541  6256.314453      False   \n",
       "79865   ENSG00000145192.12   0.0   0.0 -9.320415  4466.045703      False   \n",
       "4810    ENSG00000026025.15   0.0   0.0  9.226718  9419.225325      False   \n",
       "139314  ENSG00000197249.13   0.0   0.0 -9.106863  5051.564227      False   \n",
       "\n",
       "                grad  coef_mle   coef_sd         ll     gname  \n",
       "93003   8.236582e-07 -9.798466  0.597998 -21.964623     APOA2  \n",
       "99184   2.962665e-07 -9.657541  0.583054 -20.338973       ALB  \n",
       "79865   4.583478e-08 -9.320415  0.591606 -20.613694      AHSG  \n",
       "4810    1.069652e-06  9.226718  0.569277 -24.014993       VIM  \n",
       "139314  1.033368e-06 -9.106863  0.496072 -20.785889  SERPINA1  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(gene_ids[:5])\n",
    "gene_summary.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a name=\"det\"></a>Differential transcript expression tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly, Swan can run tests to find differentially expressed transcript isoforms. The input and output to these functions are identical to that of the differential gene tests."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Transforming to str index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "training location model: False\n",
      "training scale model: True\n",
      "iter   0: ll=292114902.725192\n",
      "iter   1: ll=292114902.725192, converged: 0.00% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter   2: ll=9689621.464688, converged: 0.64% (loc: 0.64%, scale update: True), in 418.70sec\n",
      "iter   3: ll=9689621.464688, converged: 0.64% (loc: 100.00%, scale update: False), in 0.01sec\n",
      "iter   4: ll=6067450.782106, converged: 8.86% (loc: 8.86%, scale update: True), in 358.75sec\n",
      "iter   5: ll=6067450.782106, converged: 8.86% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter   6: ll=5861723.183757, converged: 13.56% (loc: 13.56%, scale update: True), in 342.13sec\n",
      "iter   7: ll=5861723.183757, converged: 13.56% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter   8: ll=5184440.659661, converged: 15.97% (loc: 15.97%, scale update: True), in 287.90sec\n",
      "iter   9: ll=5184440.659661, converged: 15.97% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter  10: ll=5184122.283550, converged: 99.89% (loc: 99.89%, scale update: True), in 233.61sec\n",
      "iter  11: ll=5184122.283550, converged: 99.89% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter  12: ll=5183985.427572, converged: 99.97% (loc: 99.97%, scale update: True), in 7.50sec\n",
      "iter  13: ll=5183985.427572, converged: 99.97% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter  14: ll=5183973.042257, converged: 100.00% (loc: 100.00%, scale update: True), in 7.33sec\n",
      "iter  15: ll=5183973.042257, converged: 100.00% (loc: 100.00%, scale update: False), in 0.00sec\n",
      "iter  16: ll=5183973.042257, converged: 100.00% (loc: 100.00%, scale update: True), in 5.85sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/dask/array/core.py:2622: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  size = (limit / dtype.itemsize / largest_block) ** (1 / len(autos))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>pval</th>\n",
       "      <th>qval</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>mean</th>\n",
       "      <th>zero_mean</th>\n",
       "      <th>grad</th>\n",
       "      <th>coef_mle</th>\n",
       "      <th>coef_sd</th>\n",
       "      <th>ll</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21203</th>\n",
       "      <td>ENST00000367990.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.641046</td>\n",
       "      <td>6153.972461</td>\n",
       "      <td>False</td>\n",
       "      <td>2.680981e-07</td>\n",
       "      <td>-9.641046</td>\n",
       "      <td>0.599458</td>\n",
       "      <td>-21.723931</td>\n",
       "      <td>ENSG00000158874.11</td>\n",
       "      <td>APOA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6792</th>\n",
       "      <td>ENST00000295897.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.599193</td>\n",
       "      <td>5901.748828</td>\n",
       "      <td>False</td>\n",
       "      <td>7.406762e-07</td>\n",
       "      <td>-9.599193</td>\n",
       "      <td>0.583689</td>\n",
       "      <td>-20.330176</td>\n",
       "      <td>ENSG00000163631.16</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31598</th>\n",
       "      <td>ENST00000393087.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.435849</td>\n",
       "      <td>5012.441016</td>\n",
       "      <td>False</td>\n",
       "      <td>6.445624e-07</td>\n",
       "      <td>-9.435849</td>\n",
       "      <td>0.584451</td>\n",
       "      <td>-20.120285</td>\n",
       "      <td>ENSG00000197249.13</td>\n",
       "      <td>SERPINA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39289</th>\n",
       "      <td>ENST00000411641.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.236873</td>\n",
       "      <td>4108.149609</td>\n",
       "      <td>False</td>\n",
       "      <td>4.956935e-08</td>\n",
       "      <td>-9.236873</td>\n",
       "      <td>0.592742</td>\n",
       "      <td>-20.527872</td>\n",
       "      <td>ENSG00000145192.12</td>\n",
       "      <td>AHSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1044</th>\n",
       "      <td>ENST00000224237.9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.203737</td>\n",
       "      <td>9205.245637</td>\n",
       "      <td>False</td>\n",
       "      <td>1.456279e-06</td>\n",
       "      <td>9.203737</td>\n",
       "      <td>0.569203</td>\n",
       "      <td>-23.507290</td>\n",
       "      <td>ENSG00000026025.15</td>\n",
       "      <td>VIM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     tid  pval  qval    log2fc         mean  zero_mean  \\\n",
       "21203  ENST00000367990.7   0.0   0.0 -9.641046  6153.972461      False   \n",
       "6792   ENST00000295897.8   0.0   0.0 -9.599193  5901.748828      False   \n",
       "31598  ENST00000393087.8   0.0   0.0 -9.435849  5012.441016      False   \n",
       "39289  ENST00000411641.6   0.0   0.0 -9.236873  4108.149609      False   \n",
       "1044   ENST00000224237.9   0.0   0.0  9.203737  9205.245637      False   \n",
       "\n",
       "               grad  coef_mle   coef_sd         ll                 gid  \\\n",
       "21203  2.680981e-07 -9.641046  0.599458 -21.723931  ENSG00000158874.11   \n",
       "6792   7.406762e-07 -9.599193  0.583689 -20.330176  ENSG00000163631.16   \n",
       "31598  6.445624e-07 -9.435849  0.584451 -20.120285  ENSG00000197249.13   \n",
       "39289  4.956935e-08 -9.236873  0.592742 -20.527872  ENSG00000145192.12   \n",
       "1044   1.456279e-06  9.203737  0.569203 -23.507290  ENSG00000026025.15   \n",
       "\n",
       "          gname  \n",
       "21203     APOA2  \n",
       "6792        ALB  \n",
       "31598  SERPINA1  \n",
       "39289      AHSG  \n",
       "1044        VIM  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_groups = [['HepG2_1','HepG2_2'],['HFFc6_1','HFFc6_2','HFFc6_3']]\n",
    "sg.de_transcript_test(dataset_groups);\n",
    "sg.det_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And Swan can subset the results for you based on a q-value significance threshold. The resultant transcript ids can then be passed into Swan's transcript plotting function, `plot_transcript_path()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_ids, transcript_summary = sg.get_de_transcripts(q=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENST00000367990.7', 'ENST00000295897.8', 'ENST00000393087.8', 'ENST00000411641.6', 'ENST00000224237.9']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>pval</th>\n",
       "      <th>qval</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>mean</th>\n",
       "      <th>zero_mean</th>\n",
       "      <th>grad</th>\n",
       "      <th>coef_mle</th>\n",
       "      <th>coef_sd</th>\n",
       "      <th>ll</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21203</th>\n",
       "      <td>ENST00000367990.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.641046</td>\n",
       "      <td>6153.972461</td>\n",
       "      <td>False</td>\n",
       "      <td>2.680981e-07</td>\n",
       "      <td>-9.641046</td>\n",
       "      <td>0.599458</td>\n",
       "      <td>-21.723931</td>\n",
       "      <td>ENSG00000158874.11</td>\n",
       "      <td>APOA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6792</th>\n",
       "      <td>ENST00000295897.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.599193</td>\n",
       "      <td>5901.748828</td>\n",
       "      <td>False</td>\n",
       "      <td>7.406762e-07</td>\n",
       "      <td>-9.599193</td>\n",
       "      <td>0.583689</td>\n",
       "      <td>-20.330176</td>\n",
       "      <td>ENSG00000163631.16</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31598</th>\n",
       "      <td>ENST00000393087.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.435849</td>\n",
       "      <td>5012.441016</td>\n",
       "      <td>False</td>\n",
       "      <td>6.445624e-07</td>\n",
       "      <td>-9.435849</td>\n",
       "      <td>0.584451</td>\n",
       "      <td>-20.120285</td>\n",
       "      <td>ENSG00000197249.13</td>\n",
       "      <td>SERPINA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39289</th>\n",
       "      <td>ENST00000411641.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-9.236873</td>\n",
       "      <td>4108.149609</td>\n",
       "      <td>False</td>\n",
       "      <td>4.956935e-08</td>\n",
       "      <td>-9.236873</td>\n",
       "      <td>0.592742</td>\n",
       "      <td>-20.527872</td>\n",
       "      <td>ENSG00000145192.12</td>\n",
       "      <td>AHSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1044</th>\n",
       "      <td>ENST00000224237.9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.203737</td>\n",
       "      <td>9205.245637</td>\n",
       "      <td>False</td>\n",
       "      <td>1.456279e-06</td>\n",
       "      <td>9.203737</td>\n",
       "      <td>0.569203</td>\n",
       "      <td>-23.507290</td>\n",
       "      <td>ENSG00000026025.15</td>\n",
       "      <td>VIM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     tid  pval  qval    log2fc         mean  zero_mean  \\\n",
       "21203  ENST00000367990.7   0.0   0.0 -9.641046  6153.972461      False   \n",
       "6792   ENST00000295897.8   0.0   0.0 -9.599193  5901.748828      False   \n",
       "31598  ENST00000393087.8   0.0   0.0 -9.435849  5012.441016      False   \n",
       "39289  ENST00000411641.6   0.0   0.0 -9.236873  4108.149609      False   \n",
       "1044   ENST00000224237.9   0.0   0.0  9.203737  9205.245637      False   \n",
       "\n",
       "               grad  coef_mle   coef_sd         ll                 gid  \\\n",
       "21203  2.680981e-07 -9.641046  0.599458 -21.723931  ENSG00000158874.11   \n",
       "6792   7.406762e-07 -9.599193  0.583689 -20.330176  ENSG00000163631.16   \n",
       "31598  6.445624e-07 -9.435849  0.584451 -20.120285  ENSG00000197249.13   \n",
       "39289  4.956935e-08 -9.236873  0.592742 -20.527872  ENSG00000145192.12   \n",
       "1044   1.456279e-06  9.203737  0.569203 -23.507290  ENSG00000026025.15   \n",
       "\n",
       "          gname  \n",
       "21203     APOA2  \n",
       "6792        ALB  \n",
       "31598  SERPINA1  \n",
       "39289      AHSG  \n",
       "1044        VIM  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(transcript_ids[:5])\n",
    "transcript_summary.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a name=\"is\"></a>Isoform switching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We wanted to include a module to conduct rudimentary isoform switching analysis as well. We define a gene that exhibits isoform switching for our purposes as a gene that is not differentially expressed that has transcript isoforms that are differentially expressed. We have provided code to detect such instances. To run it, `de_gene_test()` and `de_transcript_test()` must first be run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_genes, is_table = sg.find_isoform_switching_genes(q=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000197746.13', 'ENSG00000067225.17', 'ENSG00000117450.13', 'ENSG00000105254.11', 'ENSG00000177600.8']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>pval</th>\n",
       "      <th>qval</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>mean</th>\n",
       "      <th>zero_mean</th>\n",
       "      <th>grad</th>\n",
       "      <th>coef_mle</th>\n",
       "      <th>coef_sd</th>\n",
       "      <th>ll</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TALONT000283514</td>\n",
       "      <td>5.099119e-10</td>\n",
       "      <td>8.195811e-08</td>\n",
       "      <td>5.538936</td>\n",
       "      <td>153.044266</td>\n",
       "      <td>False</td>\n",
       "      <td>2.116293e-07</td>\n",
       "      <td>5.538936</td>\n",
       "      <td>0.891074</td>\n",
       "      <td>-21.442096</td>\n",
       "      <td>ENSG00000197746.13</td>\n",
       "      <td>PSAP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000394936.7</td>\n",
       "      <td>9.324034e-05</td>\n",
       "      <td>5.520899e-03</td>\n",
       "      <td>-0.970022</td>\n",
       "      <td>278.640723</td>\n",
       "      <td>False</td>\n",
       "      <td>5.421814e-08</td>\n",
       "      <td>-0.970022</td>\n",
       "      <td>0.248244</td>\n",
       "      <td>-28.092922</td>\n",
       "      <td>ENSG00000197746.13</td>\n",
       "      <td>PSAP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TALONT000316712</td>\n",
       "      <td>2.490252e-09</td>\n",
       "      <td>3.622447e-07</td>\n",
       "      <td>5.038395</td>\n",
       "      <td>92.933382</td>\n",
       "      <td>False</td>\n",
       "      <td>2.948318e-08</td>\n",
       "      <td>5.038395</td>\n",
       "      <td>0.845071</td>\n",
       "      <td>-19.486628</td>\n",
       "      <td>ENSG00000067225.17</td>\n",
       "      <td>PKM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TALONT000375121</td>\n",
       "      <td>2.220446e-16</td>\n",
       "      <td>7.353454e-14</td>\n",
       "      <td>-4.855414</td>\n",
       "      <td>51.973529</td>\n",
       "      <td>False</td>\n",
       "      <td>1.056665e-08</td>\n",
       "      <td>-4.855414</td>\n",
       "      <td>0.592949</td>\n",
       "      <td>-11.909066</td>\n",
       "      <td>ENSG00000117450.13</td>\n",
       "      <td>PRDX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000585910.5</td>\n",
       "      <td>1.398881e-13</td>\n",
       "      <td>3.448465e-11</td>\n",
       "      <td>-3.687907</td>\n",
       "      <td>23.070891</td>\n",
       "      <td>False</td>\n",
       "      <td>4.299709e-01</td>\n",
       "      <td>-3.687907</td>\n",
       "      <td>0.498608</td>\n",
       "      <td>-7.304680</td>\n",
       "      <td>ENSG00000105254.11</td>\n",
       "      <td>TBCB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 tid          pval          qval    log2fc        mean  \\\n",
       "0    TALONT000283514  5.099119e-10  8.195811e-08  5.538936  153.044266   \n",
       "1  ENST00000394936.7  9.324034e-05  5.520899e-03 -0.970022  278.640723   \n",
       "2    TALONT000316712  2.490252e-09  3.622447e-07  5.038395   92.933382   \n",
       "3    TALONT000375121  2.220446e-16  7.353454e-14 -4.855414   51.973529   \n",
       "4  ENST00000585910.5  1.398881e-13  3.448465e-11 -3.687907   23.070891   \n",
       "\n",
       "   zero_mean          grad  coef_mle   coef_sd         ll                 gid  \\\n",
       "0      False  2.116293e-07  5.538936  0.891074 -21.442096  ENSG00000197746.13   \n",
       "1      False  5.421814e-08 -0.970022  0.248244 -28.092922  ENSG00000197746.13   \n",
       "2      False  2.948318e-08  5.038395  0.845071 -19.486628  ENSG00000067225.17   \n",
       "3      False  1.056665e-08 -4.855414  0.592949 -11.909066  ENSG00000117450.13   \n",
       "4      False  4.299709e-01 -3.687907  0.498608  -7.304680  ENSG00000105254.11   \n",
       "\n",
       "   gname  \n",
       "0   PSAP  \n",
       "1   PSAP  \n",
       "2    PKM  \n",
       "3  PRDX1  \n",
       "4   TBCB  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(is_genes[:5])\n",
    "is_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a name=\"es_ir\"></a>Exon skipping and intron retention"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Swan can detect novel (unannotated) exon skipping and intron retention events. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To obtain a list of genes containing novel exon skipping events, run the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing 893 intronic edges for ES\n",
      "Found 1021 novel es events from 285 genes.\n",
      "['ENSG00000130706.12', 'ENSG00000111237.18', 'ENSG00000101363.12', 'ENSG00000163069.12', 'ENSG00000132677.12']\n"
     ]
    }
   ],
   "source": [
    "es_genes = sg.find_es_genes()\n",
    "print(es_genes[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As usual, we can feed `es_genes` into `gen_report()` or individual gene ids from `es_genes` into `plot_graph()` to generate gene reports or gene summary graphs respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To obtain a list of genes containing novel intron retention events, run the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing 2185 exonic edges for IR\n",
      "Found 73 novel ir events from 47 genes.\n",
      "['ENSG00000213719.8', 'ENSG00000135480.15', 'ENSG00000196421.8', 'ENSG00000141505.11', 'ENSG00000030582.17']\n"
     ]
    }
   ],
   "source": [
    "ir_genes = sg.find_ir_genes()\n",
    "print(ir_genes[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As usual, we can feed `ir_genes` into `gen_report()` or individual gene ids from `ir_genes` into `plot_graph()` to generate gene reports or gene summary graphs respectively."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
